src/components/Editor/SymbolModal.js

```diff
@@ -0,0 +1,291 @@
+  symbolSearchResults: Array<any>,
+  selectedResultIndex: number,
+  count: number,
+  index: number
+};
+
+import "./SymbolModal.css";
+
+class SymbolModal extends Component {
+  state: SymbolModalState;
+
+  props: {
+    enabled: boolean,
+    query: string,
+    selectSource: (string, ?SelectSourceOptions) => any,
+    selectedSource?: SourceRecord,
+    symbols: SymbolDeclarations,
+    searchResults: SearchResults,
+    symbolType: SymbolSearchType,
+    setSelectedSymbolType: SymbolSearchType => any,
+    toggleActiveSearch: (?ActiveSearchType) => any,
+    setSymbolSearchQuery: string => any,
+    updateSymbolSearchResults: ({ count: number, index?: number }) => any
+  };
+
+  constructor(props) {
+    super(props);
+    this.state = {
+      symbolSearchResults: [],
+      selectedResultIndex: 0,
+      count: 0,
+      index: -1
+    };
+
+    const self = this;
+    self.onEscape = this.onEscape.bind(this);
+    self.onChange = this.onChange.bind(this);
+    self.updateResults = this.updateResults.bind(this);
+    self.traverseResults = this.traverseResults.bind(this);
+    self.renderResults = this.renderResults.bind(this);
+    self.buildSummaryMsg = this.buildSummaryMsg.bind(this);
+    self.buildPlaceHolder = this.buildPlaceHolder.bind(this);
+  }
+
+  componentWillUnmount() {
+    const shortcuts = this.context.shortcuts;
+    shortcuts.off("Escape");
+    shortcuts.off(L10N.getStr("symbolSearch.search.key2"));
+  }
+
+  componentDidMount() {
+    const shortcuts = this.context.shortcuts;
+    shortcuts.on("Escape", (_, e) => this.onEscape(e));
+    shortcuts.on(L10N.getStr("symbolSearch.search.key2"), (_, e) => {
+      this.props.toggleActiveSearch("symbol");
+      this.props.setSelectedSymbolType("functions");
+    });
+  }
+
+  componentDidUpdate(prevProps: any, prevState: any) {
+    if (this.refs.resultList && this.refs.resultList.refs) {
+      scrollList(this.refs.resultList.refs, this.state.selectedResultIndex);
+    }
+  }
+
+  onEscape(e: SyntheticEvent) {
+    e.stopPropagation();
+    e.preventDefault();
+    this.props.toggleActiveSearch();
+  }
+
+  onChange(e: SyntheticEvent) {
+    const { selectedSource, setSymbolSearchQuery } = this.props;
+    if (!selectedSource || !selectedSource.get("text")) {
+      return;
+    }
+
+    setSymbolSearchQuery(e.target.value);
+
+    return this.updateResults(e.target.value);
+  }
+
+  updateResults(query: string) {
+    const {
+      selectedSource,
+      updateSearchResults,
+      updateSymbolSearchResults,
+      selectedSymbolType,
+      symbols
+    } = this.props;
+
+    console.log(query, selectedSource);
+
+    if (query == "" || !selectedSource) {
+      return;
+    }
+
+    const symbolSearchResults = filter(symbols[selectedSymbolType], query, {
+      key: "value"
+    });
+
+    updateSearchResults({ count: symbolSearchResults.length });
+    updateSymbolSearchResults(symbolSearchResults);
+  }
+
+  traverseResults(rev: boolean) {
```

Comment:
This isn't totally necessary to change, but I found the boolean for signifying direction is a bit confusing... if I read `traversResults(false)` I expect that we are turning result traversal off. I would potentially use modulus instead to simplify things and just pass `1` or `-1`.

ie)
```js
traverseResults(direction: number) {
  const { symbolSearchResults, selectedResultIndex } = this.state;
  const searchResults = symbolSearchResults;
  const resultCount = searchResults.length;

  const updatedSelectedResult = selectedResult + direction;
  // use modulus to find the absolute fraction
  const nextResultIndex = ((number + max) % max);
  this.setState({ selectedResultIndex: nextResultIndex });
  this.onSelectResultItem(searchResults[nextResultIndex]);
}

// used like so

traverseResults(1)
traverseResults(-1)
```


If I am not mistaken, this is the second occurrence of traverse results? Should we make this a helper function that is imported?